### PR TITLE
Prevent overflow on StatsCollector.java

### PR DESCRIPTION
**Pull request by GabrielBrascher**
## Description

Did a quick check and found an operation that can result in an overflow. Additionally, I took the liberty to do a couple of code enhancements/clean ups.

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### CODE REVIEW

File: server/src/main/java/com/cloud/server/StatsCollector.java [hunk 1]
```diff
@@ -454,8 +456,9 @@ protected void init(Map<String, String> configs) {
             s_logger.warn("Usage stats job aggregation range is to small, using the minimum value of " + UsageUtils.USAGE_AGGREGATION_RANGE_MIN);
             _usageAggregationRange = UsageUtils.USAGE_AGGREGATION_RANGE_MIN;
         }
-        _diskStatsUpdateExecutor.scheduleAtFixedRate(new VmDiskStatsUpdaterTask(), (endDate - System.currentTimeMillis()), (_usageAggregationRange * 60 * 1000),
```
Comment (GabrielBrascher) [inline server/src/main/java/com/cloud/server/StatsCollector.java]: Potential overflow in 2 Values before it is converted to long by use in an invocation context.

File: server/src/main/java/com/cloud/server/StatsCollector.java [hunk 2]
```diff
@@ -1209,7 +1212,7 @@ private String getAutoscaleAction(HashMap<Long, Double> avgCounter, long groupId
             for (AutoScaleVmGroupPolicyMapVO asVmgPmap : listMap) {
                 AutoScalePolicyVO policyVO = _asPolicyDao.findById(asVmgPmap.getPolicyId());
                 if (policyVO != null) {
-                    Integer quitetime = policyVO.getQuietTime();
```
Comment (GabrielBrascher) [inline server/src/main/java/com/cloud/server/StatsCollector.java]: The variable 'quitetime' is only assigned values of primitive type and is never 'null', but it is declared with the boxed type 'Integer'.

File: server/src/main/java/com/cloud/server/StatsCollector.java [hunk 3]
```diff
@@ -1533,7 +1536,7 @@ protected void writeBatches(InfluxDB influxDbConnection, String dbName, List<Poi
      * The considered disk stats are the following: bytes read, bytes write, IO read,  and IO write.
      */
     protected boolean isCurrentVmDiskStatsDifferentFromPrevious(VmDiskStatisticsVO previousVmDiskStats, VmDiskStatisticsVO currentVmDiskStats) {
-        if (previousVmDiskStats != null) {
```
Comment (GabrielBrascher) [inline server/src/main/java/com/cloud/server/StatsCollector.java]: currentVmDiskStats might be null as well.

**Comment by GabrielBrascher** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

File: server/src/main/java/com/cloud/server/StatsCollector.java (hunk 3 above)
Comment (ustcweizhou) [reply server/src/main/java/com/cloud/server/StatsCollector.java]: @GabrielBrascher it looks currentVmDiskStats is not null


https://github.com/apache/cloudstack/blob/9a1486fdebc3d4cfd17d581ec1b66fbd9c3a4bcd/server/src/main/java/com/cloud/server/StatsCollector.java#L735-L746

Comment (GabrielBrascher) [reply server/src/main/java/com/cloud/server/StatsCollector.java]: Just updated the code addressing your comment, @ustcweizhou. Thanks for the heads up :+1:

**Comment by GabrielBrascher** [comment]
Thanks for all the reviews!

Just to mention. I did a few manual tests to ensure that no overflow will happen. Setting ONE_MINUTE_IN_MILLISCONDS to Long (or long) will indeed solve overflow issues.

**Comment by DaanHoogland** [comment]
@GabrielBrascher as this is a pure fix, will it apply to 4.13?

**Comment by GabrielBrascher** [comment]
@borisstoyanov the overflow is not something that will happen easily; however, the code is there and it still something that might happen one day depending on the value stored at `_usageAggregationRange`.

The operation that I extracted can result in an overflow as it is an **integer** operation stored in an **integer** _memory_ (and then after storing it as an integer it is cast into a long). I extracted it into a `long = int * Long`, which prevents a overflow as it is not an integer operation anymore but converted into a long operation.

**Comment by GabrielBrascher** [comment]
@DaanHoogland I see no problem in applying this at 4.13.1.

**Comment by DaanHoogland** [comment]
ok, can you rebase @GabrielBrascher ?

**Comment by GabrielBrascher** [comment]
@DaanHoogland  rebased it to 4.13 :slightly_smiling_face:

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by andrijapanicsb** [comment]
2 x LGTMs, all tests passing (those privategtw tests are "OK", fixed in newer ACS code - this PR is 150+ commits behind master)

Merging.

**Pull request merged**